File: backend/app/models/audit_log.py
from sqlalchemy import Column, String, DateTime, ForeignKey, text, func
from sqlalchemy.dialects.postgresql import UUID, JSONB, INET
from app.database import Base

class AuditLog(Base):
    __tablename__ = "audit_log"

    id = Column(UUID(as_uuid=True), primary_key=True, server_default=text("gen_random_uuid()"))
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    entity = Column(String(60), nullable=False)
    entity_id = Column(UUID(as_uuid=True), nullable=False)
    action = Column(String(30), nullable=False)
    old_value = Column(JSONB, nullable=True)
    new_value = Column(JSONB, nullable=True)
    ip_address = Column(INET, nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)

File: mobile/lib/shared/models/audit_log.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'audit_log.freezed.dart';
part 'audit_log.g.dart';

@freezed
abstract class AuditLog with _$AuditLog {
  const factory AuditLog({
    required String id,
    @JsonKey(name: 'user_id') String? userId,
    required String entity,
    @JsonKey(name: 'entity_id') required String entityId,
    required String action,
    @JsonKey(name: 'old_value') Map<String, dynamic>? oldValue,
    @JsonKey(name: 'new_value') Map<String, dynamic>? newValue,
    @JsonKey(name: 'ip_address') String? ipAddress,
    @JsonKey(name: 'created_at') required DateTime createdAt,
    String? userName,
  }) = _AuditLog;

  factory AuditLog.fromJson(Map<String, dynamic> json) => _$AuditLogFromJson(json);
}
